renormalizer/model/op.py

```diff
@@ -0,0 +1,247 @@
+        :class:`int`, :class:`str` and :class:`tuple` of them are recommended types for ``dof``.
+        For complex symbol such as ``"a^\dagger a"``, the type should be a list of hashable objects,
+        with each element representing one of the simple symbol contained in the complex symbol.
+        Using a single hashable object is also supported for complex symbol,
+        in which case every symbol is assumed to share the same DoF name.
+    factor : :class:`float` or :class:`complex` or :class:`Quantity`
+        The prefactor of the operator.
+    qn : :class:`int` or :class:`list` of :class:`int`.
+        The quantum number of the symbol. For simple symbol the ``qn`` should be an :class:`int`.
+        For complex symbol quantum number of each sinple symbol contained in the complex symbol
+        should be provided.
+
+    Notes
+    =====
+    Symbols connected by plus ``"+"`` such as ``r"b^\dagger + b"`` is considered as a simple symbol,
+    because this class is designed to deal with operator multiplication rather than addition.
+
+    Warnings
+    ========
+    If you wish to specify the DoF of each symbol in a complex symbol,
+    you should use a :class:`list` instead of :class:`tuple`.
+    Because in the latter case the :class:`tuple` is recognized as a single DoF
+    and the DoFs of all simple symbols are set to the :class:`tuple`.
+
+    """
+
+    @classmethod
+    def product(cls, op_list: List["Op"]):
+        """
+        Construct a new operator as a multiplication product of operators.
+
+        Parameters
+        ----------
+        op_list : :class:`list` of :class:`Op`
+            The operators to be multiplied.
+        Returns
+        -------
+        op : :class:`Op`
+            The product operator.
+        """
+        symbol = " ".join(op.symbol for op in op_list)
+        dof_name = list(chain.from_iterable(op.dofs for op in op_list))
+        factor = np.product([op.factor for op in op_list])
+        qn = list(chain.from_iterable(op.qn_list for op in op_list))
+        return Op(symbol, dof_name, factor, qn)
+
+    @classmethod
+    def identity(cls):
+        """
+        Construct identity operator. The DoF name is set to ``None``.
+        """
+        return cls("I", None)
+
+    def __init__(self, symbol: str, dof, factor: Union[float, Quantity] = 1.0, qn: Union[List, int] = None):
+        # This is one of the most external user interface, so detailed argument checking is necessary
+        if not isinstance(symbol, str):
+            raise TypeError(f"symbol should be a str. Got {symbol} as {type(symbol)}")
+        self.symbol: str = symbol
+        # replace " + " so that " " can be used to split the symbol ("b^\dagger + b")
+        # the logic of Op is based on multiplication of symbols. So special treatment on addition is inevitable
+        self.split_symbol : List[str] = [s.replace("plus", " + ") for s in symbol.replace(" + ", "plus").split(" ")]
```

Comment:
Do you think it is better to replace " + " in defining "b^\dagger + b" with "b^\dagger+b"?

Reply:
Well, this is a possible improvement but I usually prefer `b^\dagger + b` in writing LaTeX. In the near future, I don't think the change is necessary

Reply:
OK.